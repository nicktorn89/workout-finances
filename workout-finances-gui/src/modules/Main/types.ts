import { WorkoutObject } from 'src/store/modules/types';
import { fetchWorkouts, createWorkout, removeWorkout, changePart } from 'src/store/modules/actions';

export interface MainState {
  activeModal: boolean;
  peopleCount: number;
  isFree: boolean;
  isJumps: boolean;
  isPersonal: boolean;
  indexesToRemove: number[];
  workouts?: WorkoutObject[];
  [name: string]: boolean | number | number[] | WorkoutObject[] | undefined;
}

export interface MainProps {
  workoutsArray?: WorkoutObject[];
  fetchWorkouts?: typeof fetchWorkouts;
  createWorkout?: typeof createWorkout;
  removeWorkout?: typeof removeWorkout;
  changePart?: typeof changePart;
  currentPart?: 'first' | 'second';
  currentMonth?: number;
  currentYear?: number;
}

export interface TimeObject {
  [key: number]: {
    [key: number]: { 
      first: WorkoutObject[];
      second: WorkoutObject[];
    };
  };
}

export type ChangedMonth = {
  currentPart: 'first' | 'second',
  currentMonth: number;
  currentYear: number;
};
