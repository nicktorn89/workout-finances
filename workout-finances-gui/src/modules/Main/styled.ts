import styled from 'styled-components';
import AppBar from '@material-ui/core/AppBar';
import { Typography, Button, TextField } from '@material-ui/core';

// @ts-nocheck
export const MainHeader: any = styled(AppBar)`
  height: 50px;
`;

export const HeaderTitle: any = styled(Typography)`
  color: white !important;
  font-size: 28px !important;
  margin-left: 2rem !important;
  margin-top: 1.2rem !important;
`;

export const MainContainer = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
`;

export const ButtonsContainer = styled.div`
  width: 100%;
  height: 5rem;
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const AddWorkout: any = styled(Button)`
`;

export const PeopleNumberInput: any = styled(TextField)`
  width: 75%;
`;

export const PeopleNumberLabel = styled.label`
  width: 25%;
`;

export const SwitchLabel = styled.label`
  width: 100%;
`;
