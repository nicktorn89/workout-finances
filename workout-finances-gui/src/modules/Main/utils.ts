export const countWorkout = (workoutsCount: number) => {
  return workoutsCount * 30;
};
