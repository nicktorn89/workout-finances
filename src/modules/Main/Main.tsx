import React from 'react';
import { MainState } from './types';
import { MainHeader, HeaderTitle, MainContainer } from './styled';

import Table from 'src/components/Table';
import Slider from 'src/components/Slider';

class Main extends React.Component<{}, MainState> {
  public readonly state = {
  };

  public render = () => {
    return(
      <MainContainer>
        <MainHeader>
          <HeaderTitle 
            component='h2' 
            variant='h2'
          >
            Workout Finances
          </HeaderTitle>
        </MainHeader>

        <Slider />

        <Table />
      </MainContainer>
    );
  }
}

export default Main;
