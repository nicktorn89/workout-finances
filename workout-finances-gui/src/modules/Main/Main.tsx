import React, { ChangeEvent, MouseEvent as ReactMouseEvent } from 'react';
import { connect } from 'react-redux';
import { fetchWorkouts, createWorkout, removeWorkout } from 'src/store/modules/actions';
import { MainState, MainProps } from './types';
import {
  MainHeader, HeaderTitle, MainContainer,
  AddWorkout, ButtonsContainer, PeopleNumberInput,
  PeopleNumberLabel, SwitchLabel, RemoveWorkout,
} from './styled';
import Switch from '@material-ui/core/Switch';

import Table from 'src/components/Table';
import Slider from 'src/components/Slider';
import Modal from 'src/components/Modal';
import { countWorkout, getIdFromIndexes } from './utils';
import { RootStore } from 'src/store/types';

class Main extends React.Component<MainProps, MainState> {
  public readonly state = {
    activeModal: false,
    isPersonal: false,
    isFree: false,
    isJumps: false,
    peopleCount: 0,
    workouts: this.props.workoutsArray,
    indexesToRemove: [],
  };

  public componentDidMount = () => {
    const { fetchWorkouts } = this.props;
    fetchWorkouts && fetchWorkouts();
  }

  public componentDidUpdate = (prevProps: MainProps) => {
    if (prevProps.workoutsArray !== this.props.workoutsArray) {
      this.setState({ workouts: this.props.workoutsArray });
    }
  }

  public addIndexes = (e: ReactMouseEvent<HTMLElement, MouseEvent>) => {
    let newIndexes: number[] = [...this.state.indexesToRemove];
    const elementIndex = +(e.target as HTMLInputElement).name;

    if ((e.target as HTMLInputElement).checked) {
      newIndexes.push(elementIndex);
    } else {
      newIndexes = newIndexes.filter((index) => index !== elementIndex);
    }

    this.setState({ indexesToRemove: newIndexes });
  }

  public toggleModal = () => {
    const { activeModal } = this.state;

    this.setState({ activeModal: !activeModal });
    this.setDefaultValues();
  }

  public createWorkout = () => {
    const { peopleCount, isPersonal, isFree, isJumps } = this.state;
    const { createWorkout } = this.props;

    const workoutObject = {
      isPersonal,
      isFree,
      isJumps,
      date: new Date(),
      people: peopleCount,
      price: countWorkout(peopleCount, isPersonal, isFree, isJumps),
    };

    createWorkout && createWorkout(workoutObject);

    this.toggleModal();
  }

  public removeWorkout = () => {
    const { removeWorkout } = this.props;
    const { indexesToRemove, workouts } = this.state;

    const idArray = getIdFromIndexes(indexesToRemove, workouts!);

    removeWorkout && removeWorkout({ idArray });
    this.setState({ indexesToRemove: [] });
  }

  public setDefaultValues = () => {
    this.setState(
      {
        isFree: false,
        isJumps: false,
        isPersonal: false,
        peopleCount: 0,
      },
    );
  }

  public changePeopleCount = (e: Event) => {
    const { value } = (e.target as HTMLInputElement);
    this.setState({ peopleCount: Number(value) });
  }

  public handleSwitch = (e: ChangeEvent<HTMLInputElement>, checked: boolean) => {
    const { name } = (e.target as HTMLInputElement);
    const switches = ['isPersonal', 'isFree', 'isJumps'].filter((k) => k !== name);

    this.setState({ [name]: checked, [`${switches[0]}`]: false, [`${switches[1]}`]: false });
  }

  public render = () => {
    const { activeModal, isPersonal, isFree, isJumps, workouts } = this.state;

    return (
      <MainContainer>
        <MainHeader>
          <HeaderTitle
            component='h2'
            variant='h2'
          >
            Workout Finances
          </HeaderTitle>
        </MainHeader>

        <Slider />

        <Table
          onCheckboxChange={this.addIndexes}
          data={workouts!} 
        />

        <ButtonsContainer>
          <AddWorkout
            color='primary'
            variant='contained'
            onClick={this.toggleModal}
          >
            Добавить тренировку
          </ AddWorkout>
          
          <RemoveWorkout
            color='secondary'
            variant='contained'
            onClick={this.removeWorkout}
          >
            Удалить тренировку
          </ RemoveWorkout>
        </ButtonsContainer>

        <Modal
          isActive={activeModal}
          title='Создание записи о тренировке'
          onCancel={this.toggleModal}
          onOk={this.createWorkout}
        >
          <PeopleNumberLabel>Кол-во человек</PeopleNumberLabel>
          <PeopleNumberInput
            id='people-number'
            defaultValue={''}
            type='number'
            InputLabelProps={{
              shrink: true,
            }}
            margin='normal'
            onChange={this.changePeopleCount}
          />

          <SwitchLabel>
            <Switch
              name='isPersonal'
              onChange={this.handleSwitch}
              color='primary'
              checked={isPersonal}
            />
            Персональная тренировка
          </SwitchLabel>
          <SwitchLabel>
            <Switch
              name='isFree'
              onChange={this.handleSwitch}
              checked={isFree}
            />
            Бесплатная тренировка
          </SwitchLabel>
          <SwitchLabel>
            <Switch
              name='isJumps'
              onChange={this.handleSwitch}
              color='primary'
              checked={isJumps}
            />
            Тренировка на джампах
          </SwitchLabel>
        </Modal>
      </MainContainer>
    );
  }
}
const mapDispatchToProps = {
  fetchWorkouts,
  createWorkout,
  removeWorkout,
};

const mapStateToProps = ({ workouts }: RootStore) => ({
  workoutsArray: workouts,
});

export default connect(mapStateToProps, mapDispatchToProps)(Main);
