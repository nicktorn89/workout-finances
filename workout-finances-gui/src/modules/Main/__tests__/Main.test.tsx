import React from 'react';
import { shallow } from 'enzyme';
import Main from '../Main';
import { shallowToJson } from 'enzyme-to-json';

describe('<Main />', () => {
  const MainComponent = shallow(<Main />);
  const MainComponentToJson = shallowToJson(MainComponent);

  it('renders Main component', () => {
    expect(MainComponent).toExist();
  });

  it('should contain only one h2 title with text - Workout Finances', () => {
    expect(MainComponent).toContainExactlyOneMatchingElement('.header-title');
  });

  it('should be equal to snapshot', () => {
    expect(MainComponentToJson).toMatchSnapshot();
  });
});
