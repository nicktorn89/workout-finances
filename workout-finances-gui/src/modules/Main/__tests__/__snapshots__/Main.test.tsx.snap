// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> should be equal to snapshot 1`] = `
<styled.div>
  <Styled(WithStyles(AppBar))>
    <Styled(WithStyles(Typography))
      className="header-title"
      component="h2"
      variant="h2"
    >
      Workout Finances
    </Styled(WithStyles(Typography))>
  </Styled(WithStyles(AppBar))>
  <Slider />
  <Table />
</styled.div>
`;
